torch/_dynamo/variables/builder.py

```diff
@@ -2479,132 +2676,42 @@ def _automatic_dynamic(
-                    # yet. For case 2, we will mark the changing dimensions DYNAMIC.
-                    if not has_size_changed:
-                        for i, dim in enumerate(frame_state_entry.stride):
-                            if dim is not None and stride[i] != dim:
-                                log.debug(
-                                    "automatic dynamic %s stride(%s) %s != %s",
-                                    name,
-                                    i,
-                                    stride[i],
-                                    dim,
-                                )
-                                get_chromium_event_logger().log_instant_event(
-                                    "automatic_dynamic",
-                                    time.time_ns(),
-                                    {
-                                        "name": name,
-                                        "dim_changed": i,
-                                        "reason": "stride change",
-                                        "cached": str(dim),
-                                        "new": str(stride[i]),
-                                    },
-                                )
-                                frame_state_entry.stride[i] = None
-        tx.output.frame_state[name] = frame_state_entry
 
-    if (st := tx.distributed_state) is None:
-        stride = e.stride() if not is_sparse_any(e) else ()
-        update_frame_state(e.size(), stride)
-        frame_state_entry = tx.output.frame_state[name]
-    elif st.all_states is None:
-        # Preflight, always pretend as if it's static
-        frame_state_entry = FrameStateSizeEntry(
-            size=e.size(), scalar=None, stride=e.stride()
-        )
-        st.local_state.input_sizes[name] = list(e.size())
-        st.local_state.input_strides[name] = list(e.stride())
-    else:
-        # Apply the updates
-        for sub_state in st.all_states:
-            # Not all inputs are necessarily present on all ranks
-            if name in sub_state.input_sizes and name in sub_state.input_strides:
-                update_frame_state(
-                    sub_state.input_sizes[name], sub_state.input_strides[name]
+    # This mimics stride inference algorithm in _create_symbolic_sizes_strides_storage_offset
+    ex_size = e.size()
+    if not is_sparse_any(e):
+        ex_stride = e.stride()
+        dim = e.dim()
+
+        stride = [None] * dim
+        while any(x is None for x in stride):
+            candidates = {
+                ex_size[i] * ex_stride[i]: InferStride(i)
+                for i in range(dim)
+                if stride[i] is not None and ex_stride[i] >= 0
+            }
+            val_list = sorted(
+                [(ex_stride[i], i) for i in range(dim) if stride[i] is None],
+                key=_nested_int_aware_sort,
+            )
+            for _, i in val_list:
+                if stride[i] is None and ex_stride[i] in candidates:
```

Comment:
I think stride[i] here can't be None... (and ex_stride[i] is the ignored value from the `for` - so could use that instead of an explicit index here and on the next two lines)

Reply:
Err, why not? The loop condition is we keep looping so long as there is an entry in stride that is None.

Reply:
The ignored loop condition could be used, but I would prefer to keep it this way for symmetry with the code in symbolic_shapes.py it is modeled off of

Reply:
> Err, why not? The loop condition is we keep looping so long as there is an entry in stride that is None.

Oh - whoops. I misread the `if` test from 2447 backward.

Reply:
Actually - I take it back. I think the thing I got backwards was my original statement: How can stride[i] be anything BUT None?

The 'i' values in val_list are unique, we only add to val_list if stride[i] is None and we never assign to anything other than stride[i] in this loop (so we can't modify OTHER strides making them not None).
(In other words the test should just be: `if ex_stride[i] in candidates:`)

Reply:
i is ranging over range(dim). On subsequent iterations of the outer while loop some entries of stride will be filled and some will not.

Reply:
> i is ranging over range(dim). On subsequent iterations of the outer while loop some entries of stride will be filled and some will not.

No it's not - `i` is ranging over range(dim) *where stride[i] is None*. We recompute var_list on subsequent iterations of the outer while loop to maintain that invariant.